File: Lab14/kRUSKAL_ALGO/kRUSKAL_ALGO/kRUSKAL_ALGO.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Structure to represent an edge in the graph
struct Edge {
    int src, dest, weight;
};

// Structure to represent a disjoint-set for Union-Find
struct DisjointSet {
    int* parent, * rank;

    DisjointSet(int n) {
        parent = new int[n];
        rank = new int[n];

        // Initialize each subset as a single element set
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    // Find the set to which an element belongs (with path compression)
    int find(int u) {
        if (u != parent[u])
            parent[u] = find(parent[u]);
        return parent[u];
    }

    // Perform union of two sets (with union by rank)
    void unionSets(int u, int v) {
        int rootU = find(u);
        int rootV = find(v);

        if (rank[rootU] < rank[rootV])
            parent[rootU] = rootV;
        else if (rank[rootU] > rank[rootV])
            parent[rootV] = rootU;
        else {
            parent[rootV] = rootU;
            rank[rootU]++;
        }
    }
};
// Function to compare edges based on their weights
bool compareEdges(const Edge& a, const Edge& b) {
    return a.weight < b.weight;
}

// Function to perform Kruskal's algorithm
void kruskalMST(vector<Edge>& edges, int numVertices) {
    // Sort the edges in non-decreasing order of weights
    sort(edges.begin(), edges.end(), compareEdges);

    // Initialize a disjoint set for Union-Find
    DisjointSet disjointSet(numVertices);

    // Vector to store the edges of the MST
    vector<Edge> mst;

    for (const Edge& edge : edges) {
        int rootSrc = disjointSet.find(edge.src);
        int rootDest = disjointSet.find(edge.dest);

        // If including this edge does not create a cycle, add it to the MST
        if (rootSrc != rootDest) {
            mst.push_back(edge);
            disjointSet.unionSets(rootSrc, rootDest);
        }
    }

    // Display the edges of the Minimum Spanning Tree
    cout << "Edges of the Minimum Spanning Tree:\n";
    for (const Edge& edge : mst)
        cout << edge.src << " -- " << edge.dest << " : " << edge.weight << "\n";
}

int main() {
    // Example usage
    int numVertices = 4;
    vector<Edge> edges = { {0, 1, 10}, {0, 2, 6}, {0, 3, 5},
                           {1, 3, 15}, {2, 3, 4} };

    kruskalMST(edges, numVertices);

    return 0;
}

File: Graph/Kruskal_Algo.py
class Graph:
    def __init__(self,vertices):
        self.V=vertices
        self.graph=[]

    def addEdge(self,u,v,w):
        self.graph.append([u,v,w])
    
    def find(self,parent,i):
        if parent[i]!=i:
            parent[i]=self.find(parent,parent[i])
        return parent[i]
    
    def union(self,parent,rank,x,y):
        if rank[x]<rank[y]:
            parent[x]=y
        elif rank[x]>rank[y]:
            parent[y]=x

        else:
            parent[y]=x
            rank[x]+=1
        
    def KruskalMST(self):
        result=[]
        parent=[]
        rank=[]
        i=0
        e=0
        self.graph=sorted(self.graph, key=lambda item:item[2])

        for node in range(self.V):
            parent.append(node)
            rank.append(0)

        while e < self.V -1:
            u,v,w=self.graph[i]
            i=i+1
            x=self.find(parent,u)
            y=self.find(parent,v)

            if x!=y:
                e=e+1
                result.append([u,v,w])
                self.union(parent,rank,x,y)
        
        minCost=0
        print("Edges with minimum spanning tree:")
        for u,v,w in result:
            minCost+=w
            print(u,"--",v,"=",w)
            
        print("Minimum Spanning Tree:",minCost)

graph=Graph(4)
graph.addEdge(0,1,10)
graph.addEdge(0,2,6)
graph.addEdge(0,3,5)
graph.addEdge(1,3,15)
graph.addEdge(2,3,4)
graph.KruskalMST()

